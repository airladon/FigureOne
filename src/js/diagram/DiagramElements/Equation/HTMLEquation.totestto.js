// import HTMLEquation from './HTMLEquation';

// describe('HTML Equation', () => {
//   test('Text', () => {
//     const t = new e.Text('test', 'id1', 'c1');
//     const rendered = t.render();
//     const expected = '<div id="id1" class="equation_element c1 equation_text">\n  test\n</div>';
//     expect(rendered).toBe(expected);
//   });
//   test('Line', () => {
//     const t1 = new e.Text('a+b', 'id1', 'c1');
//     const t2 = new e.Text('=', 'id2', 'c2');
//     const t3 = new e.Text('c', 'id3', 'c3');

//     const expected =
// `<div id="id_e1" class="equation_element class_e1">
//   <div id="id1" class="equation_element c1 equation_text">
//     a+b
//   </div>
//   <div id="id2" class="equation_element c2 equation_text">
//     =
//   </div>
//   <div id="id3" class="equation_element c3 equation_text">
//     c
//   </div>
// </div>`;
//     const line = new e.Line(
//       [t1, t2, t3],
//       'id_e1',
//       'class_e1',
//     );
//     expect(line.render()).toBe(expected);
//   });

//   test('Fraction', () => {
//     const n = new e.Line(
//       [new e.Text('a + b')],
//       'id_n1',
//       'class_n1',
//     );
//     const d = new e.Line(
//       [new e.Text('c')],
//       'id_d1',
//       'class_d1',
//     );
//     const f = new e.Fraction(
//       n,
//       d,
//       'id_f1',
//       'class_f1',
//     );
//     const expected =
// `<div id="id_f1" class="equation_element class_f1 fraction">
//   <div class="numerator">
//     <div id="id_n1" class="equation_element class_n1">
//       <div class="equation_element equation_text">
//         a + b
//       </div>
//     </div>
//   </div>
//   <div class="fraction_line"> </div>
//   <div class="denominator">
//     <div id="id_d1" class="equation_element class_d1">
//       <div class="equation_element equation_text">
//         c
//       </div>
//     </div>
//   </div>
// </div>`;
//     expect(f.render()).toBe(expected);
//   });
//   describe('Equation', () => {
//     test('a + b = c', () => {
//       // const line = new e.Line();
//       const eq = new e.Equation('e1', [new e.Text('a + b = c')]);
//       const expected =
// `<div id="e1" class="equation_element equation">
//   <div class="equation_element equation_text">
//     a + b = c
//   </div>
// </div>`;
//       expect(eq.render()).toBe(expected);
//     });
//   });

//   test.only('temp', () => {
//     // console.log(e.e([e.e('b'), e.e('&#177;'), e.sqrt('4ac')]).render());
//     const n = e.e('b');
//     const d = e.e('2');
//     const eq = new e.Equation([e.frac(n, d), e.e('='), e.e('0')]);
//     console.log(eq.render());
//     expect(true).toBe(true);
//   })
// });

// // b +- sqrt(4ac)
// // -------------- = d
// //     2A

// // E = e.e([
// //   e.frac([e.e('b') e.e('+') e.sqrt('4ac')], e.e('2A')),
// //   e.('='),
// //   e.('d')])

// //  / inf
// //  |
// //  | x^2 dx = 2x
// //  |
// // / 0


// // E = e([e.o('&int;').ss('0', 'inf') e.t('x').sp('2') e.t('dx') e.('=') e.('2x')])
