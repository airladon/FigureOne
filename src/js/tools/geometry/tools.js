// @flow
import { Point } from './Point';

function pointsToNumbers2(points: Array<Point>) {
  const list = Array(points.length * 2);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 2] = points[i].x;
    list[i * 2 + 1] = points[i].y;
  }
  return list;
}

function pointsToNumbers(points: Array<Point>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return pointsToNumbers2(points);
  }
  const list = Array(points.length * 3);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 3] = points[i].x;
    list[i * 3 + 1] = points[i].y;
    list[i * 3 + 2] = points[i].z;
  }
  return list;
}

function numbersToPoints2(numbers: Array<number>) {
  const points = Array(numbers.length / 2);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(numbers[i * 2], numbers[i * 2 + 1]);
  }
  return points;
}

function numbersToPoints(numbers: Array<number>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return numbersToPoints2(numbers);
  }
  const points = Array(numbers.length / 3);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(
      numbers[i * 3],
      numbers[i * 3 + 1],
      numbers[i * 3 + 2],
    );
  }
  return points;
}

function toNumbers(
  pointsOrNumbers: Array<Point> | Array<Array<number>> | Array<number>,
  dimension: 2 | 3 = 3,
) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] === 'number') {
    return pointsOrNumbers;
  }
  if (Array.isArray(pointsOrNumbers[0])) {
    const out = [];
    for (let i = 0; i < pointsOrNumbers.length; i += 1) {
      out.push(...pointsOrNumbers);
    }
    return out;
  }
  return pointsToNumbers(pointsOrNumbers, dimension);
}

function toPoints(pointsOrNumbers: Array<Point> | Array<number>, dimension: 2 | 3 = 3) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] !== 'number') {
    return pointsOrNumbers;
  }
  return numbersToPoints(pointsOrNumbers, dimension);
}

export {
  pointsToNumbers2,
  pointsToNumbers,
  numbersToPoints,
  numbersToPoints2,
  toNumbers,
  toPoints,
};
