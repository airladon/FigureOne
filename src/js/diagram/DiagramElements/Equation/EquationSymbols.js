// @flow
import {
  Point, Transform,
} from '../../../tools/g2';
import { joinObjects } from '../../../tools/tools';
import DiagramPrimitives from '../../DiagramPrimitives/DiagramPrimitives';
// import { DiagramElementCollection } from '../../Element';
import Integral from './Symbols/Integral';
// import SuperSub from './Elements/SuperSub';
import Bracket from './Symbols/Bracket';
import Box from './Symbols/Box';
import Radical from './Symbols/Radical';
import Brace from './Symbols/Brace';
import SquareBracket from './Symbols/SquareBracket';
import AngleBracket from './Symbols/AngleBracket';
// import SquareBracketNew from './Symbols/SquareBracketNew';
import Bar from './Symbols/Bar';
import Sum from './Symbols/Sum';
import Product from './Symbols/Product';
import SimpleIntegral from './Symbols/SimpleIntegral';
import IntegralNew from './Symbols/IntegralNew';
// import BracketNew from './Symbols/BracketNew';
// import BraceNew from './Symbols/BraceNew';

// import { Annotation, AnnotationInformation } from './Elements/Annotation';
export type TypeSymbolOptions = {
  color?: Array<number>,
  numLines?: number,
  side?: 'left' | 'right' | 'bottom' | 'top',
  width?: number,
  fill?: boolean,
  staticSize?: ?(Point | [number, number] | number),
  startWidth?: number,
  lineWidth?: number,
  startHeight?: number,
  maxStartWidth?: ?number,
  maxStartHeight?: ?number,
  proportionalToHeight?: boolean,
  endLength?: number,
  sides?: number,
  tipWidth?: number,
  radius?: number,
}

export default class EquationSymbols {
  shapes: DiagramPrimitives;
  defaultColor: Array<number>;

  constructor(
    shapes: DiagramPrimitives,
    defaultColor: Array<number>,
  ) {
    this.shapes = shapes;
    this.defaultColor = defaultColor;
  }

  get(
    name: string,
    options: TypeSymbolOptions,
  ) {
    if (name === 'vinculum') {
      return this.vinculum(options);
    }
    if (name === 'strike') {
      return this.strike(options);
    }
    if (name === 'xStrike') {
      return this.xStrike(options);
    }
    if (name === 'integral') {
      return this.integral(options);
    }
    if (name === 'bracket') {         // $FlowFixMe
      return this.bracket(options);
    }
    if (name === 'squareBracket') {         // $FlowFixMe
      return this.squareBracket(options);
    }
    if (name === 'brace') {         // $FlowFixMe
      return this.brace(options);
    }
    if (name === 'bar') {         // $FlowFixMe
      return this.bar(options);
    }
    if (name === 'box') {         // $FlowFixMe
      return this.box(options);
    }
    if (name === 'angleBracket') {         // $FlowFixMe
      return this.angleBracket(options);
    }
    if (name === 'radical') {         // $FlowFixMe
      return this.radical(options);
    }
    if (name === 'simpleIntegral') {
      return this.simpleIntegral(options);
    }
    if (name === 'sum') {
      return this.sum(options);
    }
    if (name === 'prod') {
      return this.product(options);
    }
    if (name === 'int') {
      return this.integralNew(options);
    }
    return null;
  }

  vinculum(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    return this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('vinculum').scale(1, 1).translate(0, 0),
    );
  }

  box(optionsIn: {
    color?: Array<number>,
    fill?: boolean,
    width?: number,
    staticSize?: ?(Point | [number, number]),
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      fill: false,
      width: 0.01,
      staticSize: null,
    };
    const options = joinObjects(defaultOptions, optionsIn);
    return Box(this.shapes, options.color, options.fill, options.width, options.staticSize);
  }

  // simpleIntegral(optionsIn: {
  //   color?: Array<number>,
  //   lineWidth?: number,
  //   staticSize?: ?(Point | [number, number]),
  // }) {
  //   const defaultOptions: {
  //     color: this.defaultColor,
  //     lineWidth: 0.01,
  //   }
  // }

  simpleIntegral(options: {
    color?: Array<number>,
    lineWidth?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: 0.012,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new SimpleIntegral(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
      },
    )).symbol;
  }

  sum(options: {
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: null,
      sides: 5,
      staticSize: true,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new Sum(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        sides: optionsToUse.sides,
      },
    )).symbol;
  }

  product(options: {
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: null,
      sides: 5,
      staticSize: true,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new Product(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        sides: optionsToUse.sides,
      },
    )).symbol;
  }

  integralNew(options: {
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    width?: ?number,
    tipWidth?: number,
    percentage?: number,
    staticSize?: boolean,
    serif?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: null,
      width: null,
      tipWidth: null,
      percentage: 0.95,
      sides: 5,
      staticSize: true,
      radius: 0.03,
      serif: true,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new IntegralNew(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      optionsToUse,
      // {
      //   lineWidth: optionsToUse.lineWidth,
      //   minLineWidth: optionsToUse.minLineWidth,
      //   sides: optionsToUse.sides,
      //   width: optionsToUse.width,
      //   percentage: optionsToUse.percentage,
      //   sigma: optionsToUse.sigma,
      // },
    )).symbol;
  }

  radical(optionsIn: {
    color?: Array<number>,
    lineWidth?: number,
    startHeight?: number,
    startWidth?: number,
    proportionalToHeight?: boolean,
    maxStartWidth?: ?number,
    maxStartHeight?: ?number,
    staticSize?: ?(Point | [number, number]),
  }) {
    const defaultOptions: {
      color: Array<number>,
      lineWidth: number,
      startHeight: number,
      startWidth: number,
      proportionalToHeight: boolean,
      maxStartWidth: ?number,
      maxStartHeight: ?number,
      staticSize: ?(Point | [number, number]),
    } = {
      color: this.defaultColor,
      lineWidth: 0.01,
      staticSize: null,
      startHeight: 0.5,
      startWidth: 0.7,
      maxStartWidth: 0.15,
      maxStartHeight: 0.15,
      proportionalToHeight: true,
    };
    if (optionsIn.proportionalToHeight != null
      && optionsIn.proportionalToHeight === false
    ) {
      defaultOptions.startHeight = 0.15;
      defaultOptions.startWidth = 0.15;
      defaultOptions.maxStartHeight = null;
      defaultOptions.maxStartWidth = null;
    }
    const options = joinObjects(defaultOptions, optionsIn);
    return Radical(
      this.shapes, options.color, options.lineWidth,
      options.startWidth, options.startHeight, options.proportionalToHeight,
      options.maxStartWidth, options.maxStartHeight,
      options.staticSize,
    );
  }

  strike(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    return this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('strike').scale(1, 1).rotate(0).translate(0, 0),
    );
  }

  xStrike(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    const cross = this.shapes.collection(new Transform('xStrike').scale(1, 1).rotate(0).translate(0, 0));
    cross.color = color;
    const strike1 = this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('strikeLine').scale(1, 1).rotate(0).translate(0, 0),
    );
    const strike2 = strike1._dup();
    cross.add('s1', strike1);
    cross.add('s2', strike2);
    return cross;
  }

  integral(options: { color?: Array<number>, numLines?: number}) {
    const defaultOptions = {
      color: this.defaultColor,
      numLines: 1,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new Integral(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.numLines,
      new Transform('integral').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }

  bracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    width?: number,
    tipWidth?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      color: this.defaultColor,
      lineWidth: 0.012,
      sides: 10,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    if (optionsToUse.width == null) {
      optionsToUse.width = optionsToUse.lineWidth * 2.5;
    }
    if (optionsToUse.tipWidth == null) {
      optionsToUse.tipWidth = optionsToUse.lineWidth / 3;
    }
    return (new Bracket(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bracket').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        sides: optionsToUse.sides,
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
        tipWidth: optionsToUse.tipWidth,
      },
    )).symbol;
  }

  brace(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    width?: number,
    tipWidth?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      color: this.defaultColor,
      lineWidth: 0.012,
      sides: 10,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    if (optionsToUse.width == null) {
      optionsToUse.width = optionsToUse.lineWidth * 4.2;
    }
    if (optionsToUse.tipWidth == null) {
      optionsToUse.tipWidth = optionsToUse.lineWidth / 3;
    }
    return (new Brace(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        sides: optionsToUse.sides,
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
        tipWidth: optionsToUse.tipWidth,
      },
    )).symbol;
  }

  bar(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    staticSize?: ?boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      lineWidth: 0.01,
      color: this.defaultColor,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return (new Bar(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
      },
    )).symbol;
  }

  squareBracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    width?: number,
    endLineWidth?: number,
    staticSize?: ?boolean,
    radius?: number,
    sides?: number,
  }) {
    const defaultOptions = {
      side: 'left',
      lineWidth: 0.01,
      color: this.defaultColor,
      width: 0.04,
      staticSize: null,
      radius: 0,
      sides: 5,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    if (optionsToUse.endLineWidth == null) {
      optionsToUse.endLineWidth = optionsToUse.lineWidth * 0.7;
    }
    return (new SquareBracket(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
        endLineWidth: optionsToUse.endLineWidth,
        radius: optionsToUse.radius,
        sides: optionsToUse.sides,
      },
    )).symbol;
  }

  angleBracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    width?: number,
    staticSize?: ?boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      lineWidth: 0.01,
      width: 0.05,
      color: this.defaultColor,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return (new AngleBracket(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
      },
    )).symbol;
  }
}
