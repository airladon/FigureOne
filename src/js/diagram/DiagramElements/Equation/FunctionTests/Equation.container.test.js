import {
  Point,
} from '../../../../tools/g2';
import {
  round,
} from '../../../../tools/math';
import * as tools from '../../../../tools/tools';
import makeDiagram from '../../../../__mocks__/makeDiagram';
import { EquationNew } from '../Equation';

tools.isTouchDevice = jest.fn();

jest.mock('../../../Gesture');
jest.mock('../../../webgl/webgl');
jest.mock('../../../DrawContext2D');

describe('Equation Functions - Container', () => {
  let diagram;
  let eqn;
  let color1;
  let elements;
  let functions;
  // let forms;
  beforeEach(() => {
    diagram = makeDiagram();
    color1 = [0.95, 0, 0, 1];
    elements = {
      a: 'a',
      b: 'b',
      c: 'c',
    };
    functions = {
      inputForms: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const container = e.container.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Full Object
          base: {
            content: {
              container: {
                content: ['a', 'b', 'c'],
                width: null,
                descent: null,
                ascent: null,
                xAlign: 'left',
                yAlign: 'baseline',
                fit: null,
                scale: 1,
              },
            },
          },
          // Method Object
          1: {
            container: {
              content: ['a', 'b', 'c'],
              width: null,
              descent: null,
              ascent: null,
              xAlign: 'left',
              yAlign: 'baseline',
              fit: null,
              scale: 1,
            },
          },
          // Method Array
          2: {
            container: [['a', 'b', 'c'], null, null, null, 'left', 'baseline', null, 1],
          },
          // Function with Method Array
          3: e.container([['a', 'b', 'c'], null, null, null, 'left', 'baseline', null, 1]),
          // Bound Function with Object
          4: container([['a', 'b', 'c'], null, null, null, 'left', 'baseline', null, 1]),
        });
        diagram.elements = eqn;
      },
      parameterSteps: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        // const e = eqn.eqn.functions;
        // const scale = e.scale.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          base: {
            content: ['a', {
              container: {
                content: ['b'],
                width: null,
                descent: null,
                ascent: null,
                xAlign: 'left',
                yAlign: 'baseline',
                fit: null,
                scale: 1,
              },
            }, 'c'],
            scale: 1,
          },
          widthLeft: {
            content: ['a', { container: ['b', 1, null, null, 'left', 'baseline', null, 1] }, 'c'],
            scale: 1,
          },
          widthCenter: {
            content: ['a', { container: ['b', 1, null, null, 'center', 'baseline', null, 1] }, 'c'],
            scale: 1,
          },
          widthRight: {
            content: ['a', { container: ['b', 1, null, null, 'right', 'baseline', null, 1] }, 'c'],
            scale: 1,
          },
          widthMultiplier: {
            content: ['a', { container: ['b', 1, null, null, 0.2, 'baseline', null, 1] }, 'c'],
            scale: 1,
          },
          smallWidthCenter: {
            content: ['a', { container: ['b', 0.01, null, null, 'center', 'baseline', null, 1] }, 'c'],
            scale: 1,
          },
          descentAscentBaseline: {
            content: ['a', { container: ['b', null, 1, 1, 'left', 'baseline', null, 1] }, 'c'],
            scale: 1,
          },
          descentAscentBottom: {
            content: ['a', { container: ['b', null, 1, 1, 'left', 'bottom', null, 1] }, 'c'],
            scale: 1,
          },
          descentAscentMiddle: {
            content: ['a', { container: ['b', null, 1, 1, 'left', 'middle', null, 1] }, 'c'],
            scale: 1,
          },
          descentAscentTop: {
            content: ['a', { container: ['b', null, 1, 1, 'left', 'top', null, 1] }, 'c'],
            scale: 1,
          },
          descentAscentMultiplier: {
            content: ['a', { container: ['b', null, 1, 1, 'left', 0.2, null, 1] }, 'c'],
            scale: 1,
          },
          fitWidth: {
            content: ['a', { container: ['b', 1, null, null, 'left', 'baseline', 'width', 1] }, 'c'],
            scale: 1,
          },
          fitHeight: {
            content: ['a', { container: ['b', null, 1, 1, 'left', 'baseline', 'height', 1] }, 'c'],
            scale: 1,
          },
          fitContainWidth: {
            content: ['a', { container: ['b', 1, 10, 10, 'left', 'baseline', 'contain', 1] }, 'c'],
            scale: 1,
          },
          fitContainHeight: {
            content: ['a', { container: ['b', 10, 1, 1, 'left', 'baseline', 'contain', 1] }, 'c'],
            scale: 1,
          },
          scaleDefault: {
            content: ['a', { container: ['b', null, null, null, 'left', 'baseline', null, 0.5] }, 'c'],
            scale: 1,
          },
          scaleFixedContainer: {
            content: ['a', { container: ['b', 1, 1, 1, 'left', 'baseline', null, 0.5] }, 'c'],
            scale: 1,
          },
        });
        diagram.elements = eqn;
        diagram.setFirstTransform();
      },
    };
  });
  describe('Parameter Steps', () => {
    let baseA;
    let baseB;
    let baseC;
    let width;
    let smallWidth;
    let multiplier;
    let descent;
    let ascent;
    let scale;
    beforeEach(() => {
      functions.parameterSteps();
      eqn.showForm('base');
      diagram.setFirstTransform();
      baseA = eqn._a.getBoundingRect('diagram');
      baseB = eqn._b.getBoundingRect('diagram');
      baseC = eqn._c.getBoundingRect('diagram');
      width = 1;
      multiplier = 0.2;
      smallWidth = 0.01;
      descent = 1;
      ascent = 1;
      scale = 0.5;
    });
    test('Width Left', () => {
      eqn.showForm('widthLeft');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newB.left)).toEqual(round(baseA.right));
      expect(round(newC.left)).toEqual(round(baseA.right + width));
    });
    test('Width Center', () => {
      eqn.showForm('widthCenter');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newB.left)).toEqual(round(baseA.right + width / 2 - baseB.width / 2));
      expect(round(newC.left)).toEqual(round(baseA.right + width));
    });
    test('Width Right', () => {
      eqn.showForm('widthRight');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newB.left)).toEqual(round(baseA.right + width - baseB.width));
      expect(round(newC.left)).toEqual(round(baseA.right + width));
    });
    test('Width Multiplier', () => {
      eqn.showForm('widthMultiplier');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newB.left)).toEqual(round(baseA.right + width * multiplier));
      expect(round(newC.left)).toEqual(round(baseA.right + width));
    });
    test('Small Width Center', () => {
      eqn.showForm('smallWidthCenter');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newB.left)).toEqual(round(baseA.right + smallWidth / 2 - baseB.width / 2));
      expect(round(newC.left)).toEqual(round(baseA.right + smallWidth));
    });
    test('Descent Ascent Baseline', () => {
      eqn.showForm('descentAscentBaseline');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      const newC = eqn._c.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newB.left)).toEqual(round(baseB.left));
      expect(round(newC.left)).toEqual(round(baseC.left));
      expect(round(newB.bottom)).toEqual(round(baseB.bottom));
    });
    test('Descent Ascent Bottom', () => {
      eqn.showForm('descentAscentBottom');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.bottom)).toEqual(round(-descent));
    });
    test('Descent Ascent Middle', () => {
      eqn.showForm('descentAscentMiddle');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.bottom)).toEqual(round(-descent + (descent + ascent) / 2 - baseB.height / 2));
    });
    test('Descent Ascent Top', () => {
      eqn.showForm('descentAscentTop');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.top)).toEqual(round(ascent));
    });
    test('Descent Ascent Multiplier', () => {
      eqn.showForm('descentAscentMultiplier');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.bottom)).toEqual(round(-descent + (descent + ascent) * multiplier));
    });
    test('Fit Width', () => {
      eqn.showForm('fitWidth');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.width)).toEqual(round(width));
      expect(round(newB.height)).toEqual(round(width / baseB.width * baseB.height));
    });
    test('Fit Height', () => {
      eqn.showForm('fitHeight');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.height)).toEqual(round(ascent + descent));
      expect(round(newB.width)).toEqual(round((ascent + descent) / baseB.height * baseB.width));
    });
    test('Fit Contain Width', () => {
      eqn.showForm('fitContainWidth');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.width)).toEqual(round(width));
      expect(round(newB.height)).toEqual(round(width / baseB.width * baseB.height));
    });
    test('Fit Contain Height', () => {
      eqn.showForm('fitContainHeight');
      diagram.setFirstTransform();
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newB.height)).toEqual(round(ascent + descent));
      expect(round(newB.width)).toEqual(round((ascent + descent) / baseB.height * baseB.width));
    });
    test('Scale Default', () => {
      eqn.showForm('scaleDefault');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newA.width)).toEqual(round(baseA.width));
      expect(round(newB.left)).toEqual(round(baseA.right));
      expect(round(newB.width)).toEqual(round(baseB.width * scale));
    });
    test('Scale Fixed Container', () => {
      eqn.showForm('scaleFixedContainer');
      diagram.setFirstTransform();
      const newA = eqn._a.getBoundingRect('diagram');
      const newB = eqn._b.getBoundingRect('diagram');
      expect(round(newA.left)).toEqual(round(baseA.left));
      expect(round(newA.width)).toEqual(round(baseA.width));
      expect(round(newB.left)).toEqual(round(baseA.right));
      expect(round(newB.width)).toEqual(round(baseB.width * scale));
    });
  });
  test('Input Forms', () => {
    functions.inputForms();
    const elems = [eqn._a, eqn._b, eqn._c];
    const formsToTest = ['1', '2', '3', '4'];
    eqn.showForm('base');
    diagram.setFirstTransform();
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      diagram.setFirstTransform();
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Snapshot test on most simple layout
    eqn.showForm('base');
    diagram.setFirstTransform();
    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._b.transform.mat)).toMatchSnapshot();
    expect(round(eqn._c.transform.mat)).toMatchSnapshot();
  });
});
