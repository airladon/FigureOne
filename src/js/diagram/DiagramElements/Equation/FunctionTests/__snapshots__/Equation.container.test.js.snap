// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Container Input Forms 1`] = `
Array [
  0.7,
  0,
  0,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Container Input Forms 2`] = `
Array [
  0.7,
  0,
  0.07,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Container Input Forms 3`] = `
Array [
  0.7,
  0,
  0.14,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;
